File: 백준/Gold/1707. 이분 그래프/이분 그래프.js
const input = require("fs")
    .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
    .toString()
    .trim()
    .split("\n")

let idx = 0
const k = Number(input[idx++])
let result = []

while (idx < input.length) {
    const [v, e] = input[idx++].split(" ").map(Number)
    let graph = Array.from({ length: v + 1 }, () => [])
    for (let i = 0; i < e; i++) {
        const [node1, node2] = input[idx++].trim().split(" ").map(Number)
        graph[node1].push(node2)
        graph[node2].push(node1)
    }
    
    let color = Array(v + 1).fill(0)  // 0: 방문 안 함, 1/-1은 색
    let isBipartite = true

    for (let start = 1; start <= v; start++) {
        if (color[start] !== 0) continue  // 전에 방문했으면 패스

        let queue = [start]
        color[start] = 1

        for (let q = 0; q < queue.length && isBipartite; q++) {
            const cur = queue[q]

            for (let neighbor of graph[cur]) {
                if (color[neighbor] === 0) {  // 방문한 적 없으면 색 지정
                    color[neighbor] = -color[cur]
                    queue.push(neighbor)
                } else if (color[cur] === color[neighbor]) {  // 이웃한 노드끼리 색 같으면 NO
                    isBipartite = false
                    break
                }
            }
        }
        if (!isBipartite) break
    }

    result.push(isBipartite ? "YES" : "NO")
}
console.log(result.join("\n"))

File: 백준/Gold/1707. 이분 그래프/이분 그래프.java
import java.io.*;
import java.util.*;

public class Main {
	public static int k, v, e;
	public static ArrayList<Integer>[] graph;
	public static StringTokenizer st;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		k = Integer.parseInt(br.readLine());
		StringBuilder sb = new StringBuilder();

		for (int tc = 1; tc <= k; tc++) {
			st = new StringTokenizer(br.readLine());
			v = Integer.parseInt(st.nextToken());
			e = Integer.parseInt(st.nextToken());

			graph = new ArrayList[v + 1];
			for (int i = 1; i <= v; i++) {
				graph[i] = new ArrayList<>();
			}

			for (int i = 0; i < e; i++) {
				st = new StringTokenizer(br.readLine());
				int node1 = Integer.parseInt(st.nextToken());
				int node2 = Integer.parseInt(st.nextToken());
				graph[node1].add(node2);
				graph[node2].add(node1);
			}
			sb.append(solve()).append("\n");
		}
		System.out.print(sb.toString());
	}

	public static String solve() {
		int[] groups = new int[v + 1];
		for (int i = 1; i <= v; i++) {
			if (groups[i] == 0) {
				if (!check(i, groups)) {
					return "NO";
				}
			}
		}
		return "YES";
	}

	public static boolean check(int i, int[] groups) {
		Queue<Integer> q = new ArrayDeque<>();
		q.offer(i);
		groups[i] = 1;

		while (!q.isEmpty()) {
			int cur = q.poll();

			for (int neighbor : graph[cur]) { // curNode의 이웃이
				if (groups[neighbor] == 0) { // 방문한 적 없으면 무조건 반대에 넣어버리기
					groups[neighbor] = (groups[cur] == 1) ? 2 : 1;
					q.offer(neighbor);
				} else { // 이미 방문했는데 curNode랑 같은 그룹이면 땡
					if (groups[neighbor] == groups[cur]) {
						return false;
					}
				}
			}
		}
		return true;
	}
}